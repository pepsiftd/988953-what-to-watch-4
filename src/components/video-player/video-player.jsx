import React, {PureComponent, createRef} from 'react';
import PropTypes from 'prop-types';

class VideoPlayer extends PureComponent {
  constructor(props) {
    super(props);

    this._videoRef = createRef();

    this.state = {
      isMute: this.props.isMute,
      isPlaying: this.props.isPlaying,
      isLoading: true,
    };
  }

  componentDidMount() {
    const {src, poster, width, height} = this.props;
    const {isMute} = this.state;
    const video = this._videoRef.current;

    video.src = src;
    video.poster = poster;
    video.muted = isMute;
    video.width = width;
    video.height = height;

    video.oncanplaythrough = () => {
      this.setState({
        isLoading: false,
      });
    };
  }

  componentWillUnmount() {
    const video = this._videoRef.current;

    video.src = ``;
    video.poster = ``;
    video.oncanplaythrough = null;
  }

  render() {
    return (
      <video
        ref={this._videoRef}
      />
    );
  }

  componentDidUpdate() {
    const video = this._videoRef.current;

    if (this.props.isPlaying) {
      video.play();
    } else {
      video.pause();
      video.currentTime = 0;
    }
  }
}

VideoPlayer.propTypes = {
  src: PropTypes.string.isRequired,
  poster: PropTypes.string.isRequired,
  isPlaying: PropTypes.bool.isRequired,
  isMute: PropTypes.bool.isRequired,
  width: PropTypes.string.isRequired,
  height: PropTypes.string.isRequired,
};

export {VideoPlayer};
