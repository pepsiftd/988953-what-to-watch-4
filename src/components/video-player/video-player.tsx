import * as React from 'react';
import {history} from '@/history';
import {humanizeTimeElapsed} from '@/utils';

const PERCENTS = 100;

interface Props {
  children: React.ReactElement;
  isPlaying: boolean;
  title: string;
  duration: number;
  timeElapsed: number;
  onPlayClick: () => void;
  onFullscreenClick: () => void;
}

const VideoPlayer: React.FunctionComponent<Props> = ({children, isPlaying, title, duration, timeElapsed, onPlayClick, onFullscreenClick}) => {
  const timeElapsedString = humanizeTimeElapsed(timeElapsed);
  const progress = timeElapsed / duration * PERCENTS;
  return (
    <div className="player">
      {children}

      <button
        type="button"
        className="player__exit"
        onClick={() => {
          history.go(-1);
        }}
      >Exit</button>

      <div className="player__controls">
        <div className="player__controls-row">
          <div className="player__time">
            <progress className="player__progress" value={timeElapsed} max={duration}></progress>
            <div className="player__toggler" style={{left: `${progress}%`}}>Toggler</div>
          </div>
          <div className="player__time-value">{timeElapsedString}</div>
        </div>

        <div className="player__controls-row">
          {isPlaying
            ?
            <button
              type="button"
              className="player__play"
              onClick={onPlayClick}
            >
              <svg viewBox="0 0 14 21" width="14" height="21">
                <use xlinkHref="#pause"></use>
              </svg>
              <span>Pause</span>
            </button>
            :
            <button
              type="button"
              className="player__play"
              onClick={onPlayClick}
            >
              <svg viewBox="0 0 19 19" width="19" height="19">
                <use xlinkHref="#play-s"></use>
              </svg>
              <span>Play</span>
            </button>
          }

          <div className="player__name">{title}</div>

          <button
            type="button"
            className="player__full-screen"
            onClick={onFullscreenClick}
          >
            <svg viewBox="0 0 27 27" width="27" height="27">
              <use xlinkHref="#full-screen"></use>
            </svg>
            <span>Full screen</span>
          </button>
        </div>
      </div>
    </div>
  );
};

export {VideoPlayer};
