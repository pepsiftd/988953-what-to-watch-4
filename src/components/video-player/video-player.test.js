import React from 'react';
import renderer from 'react-test-renderer';

import {VideoPlayer} from './video-player';

const config = {
  src: `video-video-xxx.mp4`,
  poster: `img/video-poster-xxx.jpg`,
  isPlaying: true,
  isMute: true,
  width: `280`,
  height: `175`,
};

it(`VideoPlayer should render correctly`, () => {
  const player = renderer.create(
      <VideoPlayer
        {...config}
      />,
      {
        createNodeMock: () => {
          return {};
        }
      }
  );

  expect(player).toMatchSnapshot();
});
