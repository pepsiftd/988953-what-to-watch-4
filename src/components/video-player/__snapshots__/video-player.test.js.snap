// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer should render correctly 1`] = `<video />`;
